src/file/reader.rs

```diff
@@ -92,17 +92,54 @@ pub trait RowGroupReader {
 const FOOTER_SIZE: usize = 8;
 const PARQUET_MAGIC: [u8; 4] = [b'P', b'A', b'R', b'1'];
 
+pub trait Length  {
+  fn len(&self) -> u64;
+}
+
+pub trait TryClone: Sized {
+  fn try_clone(&self) -> Result<Self>;
+}
+
+impl Length for File {
+  fn len(&self) -> u64 {
+    self.metadata()
+        .map(|m| m.len())
+        .unwrap_or(0u64)
+  }
+}
+
+impl TryClone for File {
+  fn try_clone(&self) -> Result<Self> {
+    self.try_clone().map_err(|e| e.into())
+  }
+}
+
+impl<'a> Length for Cursor<&'a [u8]> {
+  fn len(&self) -> u64 {
+    self.get_ref().len() as u64
+  }
+}
+
+impl<'a> TryClone for Cursor<&'a [u8]> {
+  fn try_clone(&self) -> Result<Self> {
+    Ok(self.clone())
```

Comment:
I would like to point out that `try_clone` in a File returns a copy of a handler, it does not actually copy a file. In the case of Cursor, Clone would copy the content of the cursor and call `clone` on the inner buffer. So technically, these have different semantics. If this is the expected behaviour then it is okay, otherwise, we might want to revisit it.

Reply:
Note that the inner type is a reference, so cloning the cursor should clone the reference and the position in which the cursor is. If I'm not wrong, this is the same semantic as the file (so, it clones the file handler and keeps the seek position).

Reply:
Although it is true for a slice, you dont forbid any other subtypes. Semantics of TryClone should be stated clearly, when and how it should be implemented.